File: tools/utf8only.py
#!/usr/bin/env python3
"""UTF-8 is not a question: fold every test of the encoding flags.

Usage:
    python3 tools/utf8only.py <file>

Phase 12 made 'encoding' accept only utf-8, and mb_init() then sets five globals
to the same values every time: enc_utf8, has_mbyte and enc_latin1like TRUE,
enc_dbcs and enc_unicode 0.  About four hundred and fifty places still ask them,
in every shape C allows -- a bare `if`, a chain of && and ||, a ternary, a
comparison with a DBCS code page, an argument -- and each one is a branch for an
encoding this editor cannot have.

This is constant folding, done on the source, and it is careful in one way that
matters: it NEVER DROPS A SIDE EFFECT.

  1. The five globals lose their declarations and their assignments in
     mb_init(), and every other mention becomes a marker: __T__ for the three
     that are TRUE, __Z__ for the two that are 0.  A marker is an identifier, so
     the text still parses, and it is unambiguous -- a TRUE already in the
     source is never mistaken for one this tool introduced.

  2. Every expression holding a marker is simplified, innermost first, to a
     fixpoint: a ternary on a constant condition becomes its branch; in an ||
     list a false operand goes and a true one ends the list, in an && list the
     reverse; ! of a constant flips it; parentheses around a constant collapse;
     `__Z__ == DBCS_x` is false.  An operand is dropped only where C would not
     have evaluated it (after the deciding one), or where it is PURE -- no call,
     no assignment, no ++ or --.  Otherwise it stays and the expression is left
     partly unsimplified, which is correct and merely untidy.

  3. Every if, else if and while whose condition is now a constant marker folds,
     with its else chain, by brace matching.

  4. What is left -- a marker compared with something that is not a constant,
     or assigned -- becomes TRUE, FALSE or 0 again.

The sweep then takes what no longer has a caller: the DBCS and latin1 paths.
"""

import re
import sys
from pathlib import Path

sys.path.insert(0, __file__.rsplit('/', 1)[0])
import cutil

T, F, Z = '__T__', '__F__', '__Z__'
FLAGS = {'enc_utf8': T, 'has_mbyte': T, 'enc_latin1like': T, 'enc_dbcs': Z, 'enc_unicode': Z}
MARK = re.compile(r'\b(?:__T__|__F__|__Z__)\b')


def outer(s):
    """True if s is one parenthesised group, parentheses included."""
    if not s.startswith('(') or not s.endswith(')'):
        return False
    return cutil.match(s, 0, cutil.blank(s)) == len(s) - 1


def const_of(e):
    s = e.strip()
    while outer(s):
        s = s[1:-1].strip()
    if s == T:
        return True
    if s in (F, Z):
        return False
    return None


ASSIGN = re.compile(r'(?<![=!<>+\-*/%&|^])=(?!=)|\+\+|--|[+\-*/%&|^]=|<<=|>>=')
CALL = re.compile(r'[\w\]\)]\s*\(')


def pure(e):
    b = cutil.blank(e)
    return not ASSIGN.search(b) and not CALL.search(b)


def simplify(e):
    """Return e with every constant marker expression folded; spacing outside is kept."""
    lead = e[:len(e) - len(e.lstrip())]
    trail = e[len(e.rstrip()):]
    s = e.strip()
    if not s or not MARK.search(s):
        return e
    b = cutil.blank(s)
    if cutil.find_top(s, ',', 0, b) >= 0:
        return e
    # An assignment at the top of the expression is left to the statement code.
    if ASSIGN.search(''.join(ch if d == 0 else ' ' for ch, d in zip(b, cutil.depths(b)))):
        return e

    q = cutil.find_top(s, '?', 0, b)
    if q >= 0:
        cond, rest = s[:q], s[q + 1:]
        rb = cutil.blank(rest)
        colon = cutil.find_top(rest, ':', 0, rb)
        if colon < 0 or cutil.find_top(rest[:colon], '?') >= 0:
            return e
        yes, no = rest[:colon], rest[colon + 1:]
        c2 = simplify(cond)
        k = const_of(c2)
        if k is True:
            return lead + simplify(yes).strip() + trail
        if k is False:
            return lead + simplify(no).strip() + trail
        new = '%s ? %s : %s' % (c2.strip(), simplify(yes).strip(), simplify(no).strip())
        return lead + new + trail if new != s else e

    for op, stop in (('||', True), ('&&', False)):
        parts = cutil.split_top(s, op, b)
        if len(parts) > 1:
            out = []
            for p in parts:
                p2 = simplify(p).strip()
                k = const_of(p2)
                if k is (not stop):
                    continue                        # the neutral operand goes
                if k is stop:
                    if all(pure(x) for x in out):
                        return lead + (T if stop else F) + trail
                    out.append(T if stop else F)    # an impure prefix stays
                    break
                out.append(p2)
            if not out:
                return lead + (F if stop else T) + trail
            new = (' %s ' % op).join(out)
            return lead + new + trail if new != s else e

    if s.startswith('!') and not s.startswith('!='):
        inner = simplify(s[1:])
        k = const_of(inner)
        if k is True:
            return lead + F + trail
        if k is False:
            return lead + T + trail
        return lead + '!' + inner.strip() + trail if inner.strip() != s[1:].strip() else e

    if outer(s):
        inner = simplify(s[1:-1])
        k = const_of(inner)
        if k is not None:
            return lead + (T if k else F) + trail
        return lead + '(' + inner.strip() + ')' + trail if inner != s[1:-1] else e

    m = re.fullmatch(r'(__Z__|0)\s*(==|!=)\s*(__Z__|0|DBCS_\w+)', s) or \
        re.fullmatch(r'(DBCS_\w+|0)\s*(==|!=)\s*(__Z__)', s)
    if m:
        equal = (m.group(1) in (Z, '0')) and (m.group(3) in (Z, '0'))
        if m.group(2) == '==':
            return lead + (T if equal else F) + trail
        return lead + (F if equal else T) + trail
    return e


def enclosing_group(text, b, dep, pos):
    """(open, close) of the innermost parenthesis group around pos, or None."""
    d = dep[pos]
    i = pos
    while i > 0:
        i -= 1
        if b[i] == '(' and dep[i] < d:
            c = cutil.match(text, i, b)
            if c > pos:
                return i, c
            d = dep[i]
        elif b[i] in '{;}' and dep[i] < d:
            return None
    return None


def statement_bounds(text, b, dep, pos):
    """The statement around pos at brace depth: from after ; { } to its ;."""
    d = dep[pos]
    i = pos
    while i > 0 and not (b[i - 1] in ';{}' and dep[i - 1] <= d):
        i -= 1
    j = pos
    while j < len(text) and not (b[j] == ';' and dep[j] <= d):
        j += 1
    return i, j


def simplify_function(body):
    """Fold every marker expression in one function body; return (body, changes)."""
    changes = 0
    for _ in range(200):
        b = cutil.blank(body)
        dep = cutil.depths(b)
        progressed = False
        for m in MARK.finditer(body):
            pos = m.start()
            g = enclosing_group(body, b, dep, pos)
            if g:
                o, c = g
                before = body[:o].rstrip()
                control = re.search(r'\b(?:if|while|switch|for)\s*$', before) is not None
                call = (not control) and bool(before) and (before[-1].isalnum() or before[-1] in '_)]')
                inner = body[o + 1:c]
                pieces = cutil.split_top(inner, ',')
                if len(pieces) > 1 or call or re.search(r'\bfor\s*$', before):
                    seps = ';' if re.search(r'\bfor\s*$', before) else ','
                    pieces = cutil.split_top(inner, seps)
                    new_pieces = [simplify(p) for p in pieces]
                    new_inner = seps.join(new_pieces)
                    if new_inner != inner:
                        body = body[:o + 1] + new_inner + body[c:]
                        progressed = True
                        break
                    continue
                new_inner = simplify(inner)
                k = const_of(new_inner)
                if not control and k is not None and new_inner.strip() in (T, F, Z):
                    body = body[:o] + new_inner.strip() + body[c + 1:]
                    progressed = True
                    break
                if new_inner != inner:
                    body = body[:o + 1] + new_inner + body[c:]
                    progressed = True
                    break
                continue
            i, j = statement_bounds(body, b, dep, pos)
            stmt = body[i:j]
            sb = cutil.blank(stmt)
            ma = re.match(r'(\s*return\s+)', stmt)
            if ma:
                head, rhs = stmt[:ma.end()], stmt[ma.end():]
            else:
                a = re.search(r'(?<![=!<>])(?:[+\-*/%&|^]|<<|>>)?=(?!=)', sb)
                if not a:
                    continue
                head, rhs = stmt[:a.end()], stmt[a.end():]
            new_rhs = simplify(rhs)
            if new_rhs != rhs:
                body = body[:i] + head + new_rhs + body[j:]
                progressed = True
                break
        if not progressed:
            break
        changes += 1
    return body, changes


def fold_controls(body):
    """Fold if / else if / while on a constant marker; return (body, count)."""
    n = 0
    while True:
        m = re.search(r'^([ \t]*)(else if|if|while) \((__T__|__F__|__Z__)\)$', body, re.M)
        if not m:
            return body, n
        head, val = m.group(2), m.group(3)
        truth = val == T
        pat = r'^[ \t]*%s \(%s\)$' % (re.escape(head), val)
        if truth and head == 'while':
            sys.exit('utf8only: while (TRUE) from a marker -- not expected')
        if not truth:
            # A function can hold the same false condition more than once, one
            # nested inside another's block.  Folding the outer one first makes
            # the inner vanish, so fold from the LAST occurrence: its block and
            # its else chain lie after it, and nothing before it moves.
            last = list(re.finditer(pat, body, re.M))[-1]
            start = body.rfind('\n', 0, last.start()) + 1
            body = body[:start] + cutil.fold_never(body[start:], pat, 1, re.M)
            n += 1
            continue
        else:
            b = cutil.blank(body)
            mm = re.search(pat, body, re.M)
            k, o, c, _ = cutil._guarded(body, mm, b)
            end = body.index('\n', c) + 1
            tail_start = end
            while True:
                nxt = re.match(r'[ \t]*else\b', body[end:])
                if not nxt:
                    break
                o2 = b.index('{', end + nxt.end())
                c2 = cutil.match(body, o2, b)
                end = body.index('\n', c2) + 1
            if head == 'if':
                kept = cutil._dedent4(body[body.index('\n', o) + 1:body.rfind('\n', 0, c) + 1])
                body = body[:k] + kept + body[end:]
            else:
                line_end = body.index('\n', mm.start())
                body = body[:mm.start()] + m.group(1) + 'else' + body[line_end:tail_start] + body[end:]
        n += 1


def main():
    if len(sys.argv) != 2:
        sys.exit(__doc__)
    path = Path(sys.argv[1])
    t = path.read_text(errors='surrogateescape')

    t, n = re.subn(r'^static int\s+(?:enc_utf8|has_mbyte|enc_dbcs|enc_unicode|enc_latin1like)\s+=\s*[^;\n]*;\n', '', t, flags=re.M)
    if n != 5:
        sys.exit('utf8only: expected five flag declarations, found %d' % n)
    a, z = cutil.find_definition(t, 'mb_init')
    seg, n = re.subn(r'^[ \t]*(?:enc_unicode|enc_utf8|enc_dbcs|has_mbyte|enc_latin1like) = [^;\n]*;\n', '', t[a:z], flags=re.M)
    if n != 5:
        sys.exit('utf8only: expected five assignments in mb_init, found %d' % n)
    t = t[:a] + seg + t[z:]
    print('  utf8only     the five flags lose their declarations and mb_init() its assignments')

    marks = 0
    for name, tok in FLAGS.items():
        t, k = re.subn(r'\b%s\b' % name, tok, t)
        marks += k
    print('  utf8only     %d mentions become constant markers' % marks)

    import bisect
    heads = [(m.start(), m.group(1)) for m in re.finditer(r'^(\w+)\([^;\n]*\)[ \t]*\n\{', t, re.M)]
    starts = [s0 for s0, _ in heads]
    names = []
    for mk in MARK.finditer(t):
        k = bisect.bisect_right(starts, mk.start()) - 1
        if k >= 0 and heads[k][1] not in names:
            names.append(heads[k][1])
    exprs = folds = 0
    for name in names:
        span = cutil.find_definition(t, name)
        if not span:
            sys.exit('utf8only: %s holds a marker and is not a function at file scope' % name)
        a, z = span
        body, c1 = simplify_function(t[a:z])
        body, c2 = fold_controls(body)
        body, c3 = simplify_function(body)
        body, c4 = fold_controls(body)
        exprs += c1 + c3
        folds += c2 + c4
        t = t[:a] + body + t[z:]
    print('  utf8only     %d expression simplifications and %d statement folds in %d functions'
          % (exprs, folds, len(names)))

    left = MARK.findall(t)
    t = t.replace(T, 'TRUE').replace(F, 'FALSE')
    t = re.sub(r'\b__Z__\b', '0', t)
    print('  utf8only     %d constants left where they are an operand or a value, written as TRUE, FALSE or 0'
          % len(left))
    for name in FLAGS:
        if re.search(r'\b%s\b' % name, t):
            sys.exit('utf8only: %s is still named' % name)
    path.write_text(t, errors='surrogateescape')
    print('  utf8only     no test of the encoding is left to answer')


if __name__ == '__main__':
    main()

File: tools/go/internal/cut/utf8only.go
package cut

import (
	"bytes"
	"fmt"
	"io"
	"regexp"
	"sort"
	"strings"

	"slimvim.local/tools/internal/cutil"
)

const (
	markT = "__T__"
	markF = "__F__"
	markZ = "__Z__"
)

// utf8Flags are the five encoding flags and the constant each becomes.
//
// A SLICE, not a map: the substitution order decides nothing here (the five
// names are disjoint) but the count is reported, and a stable order keeps the
// two implementations saying the same thing on a refusal.
var utf8Flags = []struct{ name, tok string }{
	{"enc_utf8", markT}, {"has_mbyte", markT}, {"enc_latin1like", markT},
	{"enc_dbcs", markZ}, {"enc_unicode", markZ},
}

var (
	markRe     = regexp.MustCompile(`\b(?:__T__|__F__|__Z__)\b`)
	callRe     = regexp.MustCompile(`[\w\]\)]\s*\(`)
	returnHead = regexp.MustCompile(`^(\s*return\s+)`)
	ctrlBefore = regexp.MustCompile(`\b(?:if|while|switch|for)\s*$`)
	forBefore  = regexp.MustCompile(`\bfor\s*$`)
	elseAhead  = regexp.MustCompile(`^[ \t]*else\b`)
	ctrlMark   = regexp.MustCompile(`(?m)^([ \t]*)(else if|if|while) \((__T__|__F__|__Z__)\)$`)
	zeroCmpA   = regexp.MustCompile(`^(__Z__|0)\s*(==|!=)\s*(__Z__|0|DBCS_\w+)$`)
	zeroCmpB   = regexp.MustCompile(`^(DBCS_\w+|0)\s*(==|!=)\s*(__Z__)$`)
	funcHeads  = regexp.MustCompile(`(?m)^(\w+)\([^;\n]*\)[ \t]*\n\{`)
	flagDecls  = regexp.MustCompile(
		`(?m)^static int\s+(?:enc_utf8|has_mbyte|enc_dbcs|enc_unicode|enc_latin1like)\s+=\s*[^;\n]*;\n`)
	flagAssign = regexp.MustCompile(
		`(?m)^[ \t]*(?:enc_unicode|enc_utf8|enc_dbcs|has_mbyte|enc_latin1like) = [^;\n]*;\n`)
)

func isAssignPrefixByte(c byte) bool {
	return strings.IndexByte("+-*/%&|^", c) >= 0
}

// findAssign returns the index just past the `=` of the first assignment
// operator in b, or -1.
//
// The Python writes this as `(?<![=!<>])(?:[+\-*/%&|^]|<<|>>)?=(?!=)`, a
// NEGATIVE LOOKBEHIND, which RE2 does not have.  Spelled out: an `=` not
// followed by `=`, whose operator (with an optional +-*/%&|^ or << or >>
// before it) is not itself preceded by = ! < >.  That is what distinguishes
// `+=` and `>>=`, which ARE assignments, from `<=`, `>=`, `==` and `!=`, which
// are not.
func findAssign(b string) int {
	for i := 0; i < len(b); i++ {
		if b[i] != '=' {
			continue
		}
		if i+1 < len(b) && b[i+1] == '=' {
			continue
		}
		start := i
		if i >= 2 && (b[i-2:i] == "<<" || b[i-2:i] == ">>") {
			start = i - 2
		} else if i >= 1 && isAssignPrefixByte(b[i-1]) {
			start = i - 1
		}
		if start > 0 && strings.IndexByte("=!<>", b[start-1]) >= 0 {
			continue
		}
		return i + 1
	}
	return -1
}

// hasAssign reports whether b contains an assignment or ++ / --.
func hasAssign(b string) bool {
	if strings.Contains(b, "++") || strings.Contains(b, "--") {
		return true
	}
	return findAssign(b) >= 0
}

// outer reports whether s is one parenthesised group, parentheses included.
func outer(s string) bool {
	if !strings.HasPrefix(s, "(") || !strings.HasSuffix(s, ")") {
		return false
	}
	return cutil.Match(cutil.Blank([]byte(s)), 0) == len(s)-1
}

// constOf folds a fully-simplified expression to a constant, if it is one.
func constOf(e string) (val, ok bool) {
	s := strings.TrimSpace(e)
	for outer(s) {
		s = strings.TrimSpace(s[1 : len(s)-1])
	}
	switch s {
	case markT:
		return true, true
	case markF, markZ:
		return false, true
	}
	return false, false
}

// pure reports whether an expression has no side effect: no assignment, no
// call.  An impure operand cannot be dropped from a || or && chain.
func pure(e string) bool {
	b := string(cutil.Blank([]byte(e)))
	return !hasAssign(b) && !callRe.MatchString(b)
}

// maskDeep blanks everything at parenthesis depth above zero, so a test can
// ask about the top level only.
func maskDeep(b string) string {
	dep := cutil.Depths([]byte(b))
	out := []byte(b)
	for i := range out {
		if dep[i] != 0 {
			out[i] = ' '
		}
	}
	return string(out)
}

// simplify returns e with every constant marker expression folded; spacing
// outside is kept.
func simplify(e string) string {
	lead := e[:len(e)-len(strings.TrimLeft(e, " \t\n\r\v\f"))]
	trail := e[len(strings.TrimRight(e, " \t\n\r\v\f")):]
	s := strings.TrimSpace(e)
	if s == "" || !markRe.MatchString(s) {
		return e
	}
	sb := cutil.Blank([]byte(s))
	if cutil.FindTop([]byte(s), sb, ',') >= 0 {
		return e
	}
	// An assignment at the top of the expression is left to the statement code.
	if hasAssign(maskDeep(string(sb))) {
		return e
	}

	if q := cutil.FindTop([]byte(s), sb, '?'); q >= 0 {
		cond, rest := s[:q], s[q+1:]
		rb := cutil.Blank([]byte(rest))
		colon := cutil.FindTop([]byte(rest), rb, ':')
		if colon < 0 {
			return e
		}
		head := rest[:colon]
		if cutil.FindTop([]byte(head), cutil.Blank([]byte(head)), '?') >= 0 {
			return e
		}
		yes, no := rest[:colon], rest[colon+1:]
		c2 := simplify(cond)
		if k, ok := constOf(c2); ok {
			if k {
				return lead + strings.TrimSpace(simplify(yes)) + trail
			}
			return lead + strings.TrimSpace(simplify(no)) + trail
		}
		nw := fmt.Sprintf("%s ? %s : %s", strings.TrimSpace(c2),
			strings.TrimSpace(simplify(yes)), strings.TrimSpace(simplify(no)))
		if nw != s {
			return lead + nw + trail
		}
		return e
	}

	for _, o := range []struct {
		op   string
		stop bool
	}{{"||", true}, {"&&", false}} {
		parts := cutil.SplitTop([]byte(s), sb, o.op)
		if len(parts) > 1 {
			var out []string
			for _, p := range parts {
				p2 := strings.TrimSpace(simplify(string(p)))
				if k, ok := constOf(p2); ok {
					if k == !o.stop {
						continue // the neutral operand goes
					}
					allPure := true
					for _, x := range out {
						if !pure(x) {
							allPure = false
							break
						}
					}
					if allPure {
						if o.stop {
							return lead + markT + trail
						}
						return lead + markF + trail
					}
					if o.stop { // an impure prefix stays
						out = append(out, markT)
					} else {
						out = append(out, markF)
					}
					break
				}
				out = append(out, p2)
			}
			if len(out) == 0 {
				if o.stop {
					return lead + markF + trail
				}
				return lead + markT + trail
			}
			nw := strings.Join(out, " "+o.op+" ")
			if nw != s {
				return lead + nw + trail
			}
			return e
		}
	}

	if strings.HasPrefix(s, "!") && !strings.HasPrefix(s, "!=") {
		inner := simplify(s[1:])
		if k, ok := constOf(inner); ok {
			if k {
				return lead + markF + trail
			}
			return lead + markT + trail
		}
		if strings.TrimSpace(inner) != strings.TrimSpace(s[1:]) {
			return lead + "!" + strings.TrimSpace(inner) + trail
		}
		return e
	}

	if outer(s) {
		inner := simplify(s[1 : len(s)-1])
		if k, ok := constOf(inner); ok {
			if k {
				return lead + markT + trail
			}
			return lead + markF + trail
		}
		if inner != s[1:len(s)-1] {
			return lead + "(" + strings.TrimSpace(inner) + ")" + trail
		}
		return e
	}

	m := zeroCmpA.FindStringSubmatch(s)
	if m == nil {
		m = zeroCmpB.FindStringSubmatch(s)
	}
	if m != nil {
		isZero := func(x string) bool { return x == markZ || x == "0" }
		equal := isZero(m[1]) && isZero(m[3])
		if m[2] == "==" {
			if equal {
				return lead + markT + trail
			}
			return lead + markF + trail
		}
		if equal {
			return lead + markF + trail
		}
		return lead + markT + trail
	}
	return e
}

// enclosingGroup returns the innermost parenthesis group around pos.
func enclosingGroup(text string, b []byte, dep []int, pos int) (int, int, bool) {
	d := dep[pos]
	i := pos
	for i > 0 {
		i--
		if b[i] == '(' && dep[i] < d {
			c := cutil.Match(b, i)
			if c > pos {
				return i, c, true
			}
			d = dep[i]
		} else if (b[i] == '{' || b[i] == ';' || b[i] == '}') && dep[i] < d {
			return 0, 0, false
		}
	}
	return 0, 0, false
}

// statementBounds gives the statement around pos at brace depth: from after
// ; { } to its ;.
func statementBounds(text string, b []byte, dep []int, pos int) (int, int) {
	d := dep[pos]
	i := pos
	for i > 0 && !(strings.IndexByte(";{}", b[i-1]) >= 0 && dep[i-1] <= d) {
		i--
	}
	j := pos
	for j < len(text) && !(b[j] == ';' && dep[j] <= d) {
		j++
	}
	return i, j
}

// simplifyFunction folds every marker expression in one function body.
func simplifyFunction(body string) (string, int) {
	changes := 0
	for round := 0; round < 200; round++ {
		b := cutil.Blank([]byte(body))
		dep := cutil.Depths(b)
		progressed := false
		for _, mk := range markRe.FindAllStringIndex(body, -1) {
			pos := mk[0]
			if o, c, ok := enclosingGroup(body, b, dep, pos); ok {
				before := strings.TrimRight(body[:o], " \t\n\r\v\f")
				control := ctrlBefore.MatchString(before)
				call := !control && before != "" &&
					(isAlnumByte(before[len(before)-1]) || before[len(before)-1] == '_' ||
						before[len(before)-1] == ')' || before[len(before)-1] == ']')
				inner := body[o+1 : c]
				ib := cutil.Blank([]byte(inner))
				pieces := cutil.SplitTop([]byte(inner), ib, ",")
				isFor := forBefore.MatchString(before)
				if len(pieces) > 1 || call || isFor {
					sep := ","
					if isFor {
						sep = ";"
					}
					pieces = cutil.SplitTop([]byte(inner), ib, sep)
					newPieces := make([]string, len(pieces))
					for i, p := range pieces {
						newPieces[i] = simplify(string(p))
					}
					newInner := strings.Join(newPieces, sep)
					if newInner != inner {
						body = body[:o+1] + newInner + body[c:]
						progressed = true
						break
					}
					continue
				}
				newInner := simplify(inner)
				_, isConst := constOf(newInner)
				t := strings.TrimSpace(newInner)
				if !control && isConst && (t == markT || t == markF || t == markZ) {
					body = body[:o] + t + body[c+1:]
					progressed = true
					break
				}
				if newInner != inner {
					body = body[:o+1] + newInner + body[c:]
					progressed = true
					break
				}
				continue
			}
			i, j := statementBounds(body, b, dep, pos)
			stmt := body[i:j]
			sb := string(cutil.Blank([]byte(stmt)))
			var head, rhs string
			if ma := returnHead.FindStringIndex(stmt); ma != nil {
				head, rhs = stmt[:ma[1]], stmt[ma[1]:]
			} else {
				a := findAssign(sb)
				if a < 0 {
					continue
				}
				head, rhs = stmt[:a], stmt[a:]
			}
			newRhs := simplify(rhs)
			if newRhs != rhs {
				body = body[:i] + head + newRhs + body[j:]
				progressed = true
				break
			}
		}
		if !progressed {
			break
		}
		changes++
	}
	return body, changes
}

func isAlnumByte(c byte) bool {
	return ('0' <= c && c <= '9') || ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z')
}

// foldControls folds if / else if / while on a constant marker.
func foldControls(body string) (string, int, error) {
	n := 0
	for {
		m := ctrlMark.FindStringSubmatchIndex(body)
		if m == nil {
			return body, n, nil
		}
		indent := body[m[2]:m[3]]
		head := body[m[4]:m[5]]
		val := body[m[6]:m[7]]
		truth := val == markT
		pat := regexp.MustCompile(`(?m)^[ \t]*` + regexp.QuoteMeta(head) + ` \(` + val + `\)$`)
		if truth && head == "while" {
			return "", 0, fmt.Errorf("utf8only: while (TRUE) from a marker -- not expected")
		}
		if !truth {
			// A function can hold the same false condition more than once, one
			// nested inside another's block.  Folding the OUTER one first makes
			// the inner vanish, so fold from the LAST occurrence: its block and
			// its else chain lie after it, and nothing before it moves.
			all := pat.FindAllStringIndex(body, -1)
			last := all[len(all)-1]
			start := strings.LastIndexByte(body[:last[0]], '\n') + 1
			out, err := cutil.FoldNever([]byte(body[start:]), pat.String(), 1)
			if err != nil {
				return "", 0, fmt.Errorf("utf8only: %v", err)
			}
			body = body[:start] + string(out)
			n++
			continue
		}
		b := cutil.Blank([]byte(body))
		mm := pat.FindStringIndex(body)
		k, o, c, _, err := cutil.Guarded([]byte(body), b, mm)
		if err != nil {
			return "", 0, fmt.Errorf("utf8only: %v", err)
		}
		end := c + strings.IndexByte(body[c:], '\n') + 1
		tailStart := end
		for {
			nxt := elseAhead.FindStringIndex(body[end:])
			if nxt == nil {
				break
			}
			at := end + nxt[1]
			o2 := at + bytes.IndexByte(b[at:], '{')
			c2 := cutil.Match(b, o2)
			if c2 < 0 {
				return "", 0, fmt.Errorf("utf8only: an else block is unbalanced")
			}
			end = c2 + strings.IndexByte(body[c2:], '\n') + 1
		}
		if head == "if" {
			kept := string(cutil.Dedent4([]byte(
				body[o+strings.IndexByte(body[o:], '\n')+1 : strings.LastIndexByte(body[:c], '\n')+1])))
			body = body[:k] + kept + body[end:]
		} else {
			lineEnd := mm[0] + strings.IndexByte(body[mm[0]:], '\n')
			body = body[:mm[0]] + indent + "else" + body[lineEnd:tailStart] + body[end:]
		}
		n++
	}
}

// Utf8Only leaves no test of the encoding to answer.
func Utf8Only(text []byte, w io.Writer) ([]byte, error) {
	if n := len(flagDecls.FindAll(text, -1)); n != 5 {
		return nil, fmt.Errorf("utf8only: expected five flag declarations, found %d", n)
	}
	text = flagDecls.ReplaceAll(text, nil)

	a, z, ok := cutil.FindDefinition(text, cutil.Blank(text), "mb_init")
	if !ok {
		return nil, fmt.Errorf("utf8only: mb_init is not defined at file scope")
	}
	seg := text[a:z]
	if n := len(flagAssign.FindAll(seg, -1)); n != 5 {
		return nil, fmt.Errorf("utf8only: expected five assignments in mb_init, found %d", n)
	}
	seg = flagAssign.ReplaceAll(seg, nil)
	var buf []byte
	buf = append(buf, text[:a]...)
	buf = append(buf, seg...)
	text = append(buf, text[z:]...)
	fmt.Fprintln(w, "  utf8only     the five flags lose their declarations and mb_init() "+
		"its assignments")

	marks := 0
	for _, f := range utf8Flags {
		re := regexp.MustCompile(`\b` + f.name + `\b`)
		marks += len(re.FindAll(text, -1))
		text = re.ReplaceAll(text, []byte(f.tok))
	}
	fmt.Fprintf(w, "  utf8only     %d mentions become constant markers\n", marks)

	// Which functions hold a marker, in the order they first appear -- the
	// Python builds this with bisect over the function heads.
	type head struct {
		at   int
		name string
	}
	var heads []head
	for _, m := range funcHeads.FindAllSubmatchIndex(text, -1) {
		heads = append(heads, head{m[0], string(text[m[2]:m[3]])})
	}
	starts := make([]int, len(heads))
	for i, h := range heads {
		starts[i] = h.at
	}
	var names []string
	seen := map[string]bool{}
	for _, mk := range markRe.FindAllIndex(text, -1) {
		k := sort.SearchInts(starts, mk[0]+1) - 1
		if k >= 0 && !seen[heads[k].name] {
			seen[heads[k].name] = true
			names = append(names, heads[k].name)
		}
	}

	exprs, folds := 0, 0
	for _, name := range names {
		fa, fz, ok := cutil.FindDefinition(text, cutil.Blank(text), name)
		if !ok {
			return nil, fmt.Errorf("utf8only: %s holds a marker and is not a function at "+
				"file scope", name)
		}
		body := string(text[fa:fz])
		body, c1 := simplifyFunction(body)
		body, c2, err := foldControls(body)
		if err != nil {
			return nil, err
		}
		body, c3 := simplifyFunction(body)
		body, c4, err := foldControls(body)
		if err != nil {
			return nil, err
		}
		exprs += c1 + c3
		folds += c2 + c4
		var nb []byte
		nb = append(nb, text[:fa]...)
		nb = append(nb, body...)
		text = append(nb, text[fz:]...)
	}
	fmt.Fprintf(w, "  utf8only     %d expression simplifications and %d statement folds "+
		"in %d functions\n", exprs, folds, len(names))

	left := len(markRe.FindAll(text, -1))
	text = bytes.ReplaceAll(text, []byte(markT), []byte("TRUE"))
	text = bytes.ReplaceAll(text, []byte(markF), []byte("FALSE"))
	text = regexp.MustCompile(`\b__Z__\b`).ReplaceAll(text, []byte("0"))
	fmt.Fprintf(w, "  utf8only     %d constants left where they are an operand or a value, "+
		"written as TRUE, FALSE or 0\n", left)
	for _, f := range utf8Flags {
		if regexp.MustCompile(`\b` + f.name + `\b`).Match(text) {
			return nil, fmt.Errorf("utf8only: %s is still named", f.name)
		}
	}
	fmt.Fprintln(w, "  utf8only     no test of the encoding is left to answer")
	return text, nil
}
